pkg/flypg/node.go

```diff
@@ -234,10 +270,98 @@ func (n *Node) PostInit(ctx context.Context) error {
 		if member != nil && member.Role != "" {
 			role = member.Role
 		}
+		fmt.Printf("My current role is: %s\n", role)
 
 		switch role {
 		case PrimaryRoleName:
-			// TODO - This is where we need to fence the primary if it's no longer valid.
+			standbys, err := repmgr.StandbyMembers(ctx, conn)
+			if err != nil {
+				if !errors.Is(err, pgx.ErrNoRows) {
+					return fmt.Errorf("failed to query standbys")
+				}
+			}
+
+			totalMembers := len(standbys) + 1 // include self
+			totalActive := 1                  // include self
+			totalInactive := 0
+			totalConflicts := 0
+			conflictMap := map[string]int{}
+
+			for _, standby := range standbys {
+				mConn, err := repmgr.NewRemoteConnection(ctx, standby.Hostname)
+				if err != nil {
+					fmt.Printf("failed to connect to %s", standby.Hostname)
+					totalInactive++
+					continue
+				}
+
+				primary, err := repmgr.PrimaryMember(ctx, mConn)
+				if err != nil {
+					fmt.Printf("failed to resolve primary from standby %s", standby.Hostname)
+					totalInactive++
+					continue
+				}
+
+				totalActive++
+
+				if primary.Hostname != n.PrivateIP {
+					totalConflicts++
+					conflictMap[primary.Hostname]++
+				}
+			}
+
+			// Using the cluster state metrics, determine whether it's safe to boot as a primary.
+			primary, err := ZombieDiagnosis(n.PrivateIP, totalMembers, totalInactive, totalActive, conflictMap)
+			if err != nil {
+				if errors.Is(err, ErrZombieDiscovered) {
+					fmt.Println("Unable to confirm we are the real primary!")
+					fmt.Printf("Registered members: %d, Active member(s): %d, Inactive member(s): %d, Conflicts detected: %d\n",
+						totalMembers,
+						totalActive,
+						totalInactive,
+						totalConflicts,
+					)
+
+					fmt.Println("Identifying ourself as a Zombie")
+
+					// if primary is non-empty we were able to build a consensus on who the real primary and should
+					// be recoverable on reboot.
+					if primary != "" {
+						fmt.Printf("Majority of members agree that %s is the real primary\n", primary)
+						fmt.Println("Reconfiguring PGBouncer to point to the real primary")
+						if err := n.PGBouncer.ConfigurePrimary(ctx, primary, true); err != nil {
+							return fmt.Errorf("failed to reconfigure pgbouncer: %s", err)
+						}
+					}
+					// Create a zombie.lock file containing the resolved primary.
+					// This will be an empty string if we are unable to resolve the real primary.
+					if err := writeZombieLock(primary); err != nil {
+						return fmt.Errorf("failed to set zombie lock: %s", err)
+					}
+
+					fmt.Println("Setting all existing tables to read-only")
+					if err := admin.SetReadOnly(ctx, conn); err != nil {
+						return fmt.Errorf("failed to set read-only: %s", err)
+					}
+
+					return fmt.Errorf("zombie primary detected. Use `fly machines restart <machine-id>` to rejoin the cluster or consider removing this node")
+				}
+
+				return fmt.Errorf("failed to run zombie diagnosis: %s", err)
+			}
```

Comment:
You can reduce nesting a bit by moving the `errors.Is()` outside the outer `if`:

```go
if errors.Is(err, ErrZombieDiscovered) {
	...
} else if err != nil {
	...
}
```